torch/ao/quantization/fx/prepare.py

```diff
@@ -1295,10 +1305,26 @@ def insert_observers_for_model(
                     is_reuse_input_qconfig_ = is_reuse_input_qconfig(qconfig)
 
                     if is_last_node_of_pattern:
-                        # this returns the new observer node if it was needed
-                        maybe_output_obs_node = maybe_insert_output_observer_for_node(
-                            node, model, modules, graph, matches,
-                            node_name_to_target_dtype, pattern, qhandler, is_qat)
+                        maybe_output_obs_node = None
+                        if _is_custom_module_lstm(node, modules, qconfig, qhandler):
+                            # Currently custom module outputs are assumed to be already quantized,
+                            # so we need to insert a DeQuantStub after the output. For custom module
+                            # LSTM specifically, the outputs are also a nested tuple, so we must first
+                            # break down the tuple to insert DeQuantStubs after the internal nodes.
+
+                            # TODO: This currently diverges from how custom modules are handled today,
+                            # where we insert observers after the output instead of DeQuantStubs, and
+                            # replace these observers with "dequantize" nodes during convert. Conceptually,
+                            # these output observers are the same as DeQuantStubs. In the future, we
+                            # should resolve this inconsistency by inserting DeQuantStubs for all custom
+                            # modules, not just for LSTM.
+                            _insert_dequant_stubs_for_custom_module_lstm_output(node, model, modules, graph)
```

Comment:
it feels like `_insert_dequant_stubs_for_custom_module_lstm_output` is not doing equivalent things as `maybe_insert_output_observer_for_node`, e.g. `maybe_output_obs_node` will go through L1328-1348 to update the user node to use this, but looks like this is done inside `_insert_dequant_stubs_for_custom_module_lstm_output`
I'm wondering if branching is placed in a wrong place, should the branch be one level outer?

Reply:
I agree. The only code that we still need for the LSTM case is [these two lines](https://github.com/pytorch/pytorch/blob/4092c24e234ebb2c4ff5eb40b4eb3ebe187003c2/torch/ao/quantization/fx/prepare.py#L1361-L1362):
```
if qhandler is not None and qhandler.is_custom_module():
    swap_custom_module_to_observed(node, qconfig, modules, prepare_custom_config)
```
Maybe I should just copy the second line (the swap) to the LSTM case?

Reply:
Ok, I copied that line over to the LSTM case and put everything else in the else case